### PR TITLE
Add test plan for QUARKUS-4430 | http metrics should provide a path instead of REDIRECTION and NOT_FOUND

### PR DESCRIPTION
**Pull request by jcarranzan**
### Links

JIRA:
https://issues.redhat.com/browse/QUARKUS-4430

### Reminder for considerable topics

 - [x] Make sure you have considered the following areas when preparing the test plan:

   - Logging
   - Tracing
   - Metrics
   - Security
   - OpenAPI
   - Data sources
   - Frontend
   - Qute

### CODE REVIEW

File: QUARKUS-4430.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+* We have already tests related to monitoring metrics with Prometheus and OpenId Connect Keycloak but not specific test
+  about this improvement feature related to tag redirections and not found responses.
+ 
+
+### Impact on test suites and testing automation
+* New test class will be added into [monitoring](https://github.com/quarkus-qe/quarkus-test-suite/tree/main/monitoring/micrometer-prometheus-oidc) 
+
+### Impact on resources
+* One baremetal test execution in JVM mode
+* Test require any KeyCloak service container
```
Comment (michalvavrik) [inline QUARKUS-4430.md:28]:
```suggestion
* Test require any Keycloak service container
```
I don't understand why it would require Keycloak at all, can you elaborate?

File: QUARKUS-4430.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+## Existing test coverage
+* We have already tests related to monitoring metrics with Prometheus and OpenId Connect Keycloak but not specific test
+  about this improvement feature related to tag redirections and not found responses.
+ 
+
+### Impact on test suites and testing automation
+* New test class will be added into [monitoring](https://github.com/quarkus-qe/quarkus-test-suite/tree/main/monitoring/micrometer-prometheus-oidc) 
+
+### Impact on resources
+* One baremetal test execution in JVM mode
```
Comment (michalvavrik) [inline QUARKUS-4430.md:27]: we generally test everything in both JVM and native mode unless we have a reason for the opposite; now, the Keycloak only runs in JVM IIRC but we don't care about Keycloak per say, just about Quarkus

File: QUARKUS-4430.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+* Test require any KeyCloak service container
+* Overall QE Test Suite execution time will be greater roughly by 10 seconds
```
Comment (michalvavrik) [inline QUARKUS-4430.md:29]: that can only be true if you integrate new test coverage into existing test class, can you do that? if so, cool

File: QUARKUS-4430.md [hunk 4]
```diff
@@ -0,0 +1,35 @@
+# QUARKUS-3535 - Add a metric for rejected submissions to the pool in RHBQ 3.2
+
+JIRA: https://issues.redhat.com/browse/QUARKUS-4430
+
+PR: https://github.com/quarkusio/quarkus/pull/39583
+
+Upstream issue: https://github.com/quarkusio/quarkus/issues/39581
+
+Verify that URIs in metrics tags are correctly recorded when they match a client or server pattern,
```
Comment (michalvavrik) [inline QUARKUS-4430.md:9]: my understanding is that this change is about path templates, so having test coverage for 302 and 404 is cool, but please stress in the TP what is it that you want to test, e.g. `/moved/{id}`. because you need to create a test coverage that pass with 3.8.5 and but fails with 3.8.4. thanks

now, I understand that this text comes from the PR title `Keep the URIs in the metrics tag if they match a client or server pattern`, but I am not convinced it's correct term. It is path pattern in either client or server path.

File: QUARKUS-4430.md [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+especially in cases of redirection (302) and not found (404) responses.
+
+## Scope of the testing
+* Run Quarkus micrometer Prometheus application with OpenId Connect Keycloak and verify in case of redirections (302)
```
Comment (michalvavrik) [inline QUARKUS-4430.md:13]: _OpenId Connect Keycloak_ is not really relevant for scope of testing, it's relevant for a motivation of this change

File: QUARKUS-4430.md [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+  and not found (404) response the Uris metrics tags are correctly recorded.
+* The test coverage will be verified for JVM.
```
Comment (michalvavrik) [inline QUARKUS-4430.md:15]: judging by code changes this will work in native, but if you integrate this into existing test class, you will run it in native anyway (as you should)

File: QUARKUS-4430.md [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+JIRA: https://issues.redhat.com/browse/QUARKUS-4430
+
+PR: https://github.com/quarkusio/quarkus/pull/39583
+
+Upstream issue: https://github.com/quarkusio/quarkus/issues/39581
+
+Verify that URIs in metrics tags are correctly recorded when they match a client or server pattern,
+especially in cases of redirection (302) and not found (404) responses.
+
+## Scope of the testing
```
Comment (michalvavrik) [inline QUARKUS-4430.md:12]: I'd like you to add test cases for paths with empty path segments; the https://github.com/quarkusio/quarkus/pull/39583/files is comparing paths and we should be sure that it is using normalized paths (which I suppose it does, but let's have tests). e.g. `/moved////{id}`

**Review by michalvavrik — COMMENTED** [review]
TP looks fine, but I still found a way to complain; please see comments below

File: QUARKUS-4430.md (hunk 1 above)
Comment (jcarranzan) [reply QUARKUS-4430.md:28]: Well, this issue development has been done for a Keycloak request, you can see here--> https://issues.redhat.com/browse/QUARKUS-4430 
_"Keycloak wants to raise a request to backport the enhancement for reflecting redirected URLs on HTTP Metrics. The Quarkus version in question is 3.8.

The relevant PR to be backported: https://github.com/quarkusio/quarkus/pull/39583 "_

Comment (michalvavrik) [reply QUARKUS-4430.md:28]: Let's add it to the motivation (the description paragraph), but you cannot use Keycloak in this testing because you don't have a use case. If you have in mind how to use the Keycloak in relation to the backport, please show me.

File: QUARKUS-4430.md (hunk 3 above)
Comment (jcarranzan) [reply QUARKUS-4430.md:29]: I'll try it.

File: QUARKUS-4430.md (hunk 1 above)
Comment (jcarranzan) [reply QUARKUS-4430.md:28]: For that reason, my idea was create the tests into https://github.com/quarkus-qe/quarkus-test-suite/blob/main/monitoring/micrometer-prometheus-oidc/ that it's used Keycloak and prometheus metric, so in that case we will have an integration test checking those urls for the cases Not found and redirection....

Comment (jcarranzan) [reply QUARKUS-4430.md]: I've moved the test just  to /monitoring/micrometer-prometheus without keycloak because is not relevant here.

File: QUARKUS-4430.md [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+* Run Quarkus micrometer Prometheus application and verify metric tags for:
+   * Redirections (HTTP 302)
+   *   Not Found responses (HTTP 404)
```
Comment (michalvavrik) [inline QUARKUS-4430.md:15]:
```suggestion
   * Not Found responses (HTTP 404)
```

File: QUARKUS-4430.md [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+   *  Empty path responses (HTTP 204)
+   *  Dynamic path segments (HTTP 301)
```
[the remark this answers is not in the document]
Comment (michalvavrik) [reply QUARKUS-4430.md:17]: BTW why are they called dynamic path segments? I am just curious why you use this word for 301. I'm probably missing something.

**Review by michalvavrik — APPROVED** [review]
